File: diag_litellm.py
#!/usr/bin/env python3
"""Diagnostic: Find qwen3-6-nvfp4 to internal UUID mapping in LiteLLM proxy."""
import sys, json, urllib.request, http.client, ssl

BASE = sys.argv[1] if len(sys.argv) > 1 else ""
KEY = sys.argv[2] if len(sys.argv) > 2 else ""
SLUG = "qwen3-6-nvfp4"

def make_request(path):
    from urllib.parse import urlparse
    parsed = urlparse(BASE.rstrip("/") + path)
    
    ctx = ssl.create_default_context()
    
    conn = http.client.HTTPSConnection(parsed.hostname, timeout=15) if parsed.scheme == "https" else None
    if conn is None:
        conn = http.client.HTTPConnection(parsed.hostname, timeout=15)
    
    headers = {
        "Authorization": f"Bearer {KEY}",
        "Content-Type": "application/json",
    }
    
    try:
        conn.request("GET", parsed.path, headers=headers)
        resp = conn.getresponse()
        body = resp.read().decode()
        code = resp.status
        conn.close()
        return code, body
    except Exception as e:
        conn.close()
        return 0, str(e)

print(f"# SLUG={SLUG}")

# A: GET /models (OpenAI compatible)
code, body = make_request("/models")
print(f"\n--- GET /models HTTP:{code} size:{len(body)} ---")
try:
    d = json.loads(body)
    items = d.get("data", [])
    print(f"Total: {len(items)} models")
    # Show entries with slug
    for m in items:
        s = json.dumps(m).lower()
        if "qwen" in s.lower():
            print(json.dumps(m, indent=2))
except Exception as e:
    print(f"Parse error: {e}")
    print(body[:500])

# B: current broken approach
code, body = make_request(f"/model/info?qwen3-6-nvfp4")
print(f"\n--- GET /model/info?qwen3-6-nvfp4 HTTP:{code} ---")
print(body[:500])

# C: ALL models via /model/info
code, body = make_request("/model/info")
print(f"\n--- GET /model/info all HTTP:{code} size:{len(body)} ---")
if code == 200:
    d = json.loads(body)
    mi = d.get("model_info", {})
    print(f"Keyed models: {len(mi)}")
    
    found = None
    for k, v in mi.items():
        kl = k.lower()
        tn = str(v.get("model_name", "")).lower()
        if SLUG.lower() in kl or SLUG.lower() in tn:
            mid = v.get("model_info", {}).get("id", "NONE")
            nm = v.get("model_name", "")
            found = (k, nm, mid)
            break
    
    if found:
        key, name, uid = found
        print(f"\n>>> FOUND: key='{key}' model_name='{name}' id='{uid}'")
    else:
        print(f"\nNO MATCH for '{SLUG}'. Sample keys:")
        for i, k in enumerate(list(mi.keys())[:8]):
            nm = mi[k].get("model_name", "")
            uid = mi[k].get("model_info", {}).get("id", "<none>")
            print(f"  #{i}: '{k}' name='{nm}' id='{uid}'")
        
        # show model_name values specifically  
        names = set()
        for v in mi.values():
            n = v.get("model_name", "")
            if n: names.add(n)
        print(f"\nDistinct model_names ({len(names)}):")
        for n in sorted(names)[:20]:
            print(f"  {n!r}")

File: diag_litellm.go
// +build ignore

package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

func main() {
	if len(os.Args) < 3 {
		fmt.Fprintf(os.Stderr, "Usage: %s <LITELLM_BASE_URL> <API_KEY>\n", os.Args[0])
		os.Exit(1)
	}
	baseURL := strings.TrimRight(os.Args[1], "/")
	apiKey := os.Args[2]
	slug := "qwen3-6-nvfp4"

	c := &http.Client{Timeout: 30 * time.Second}

	fmt.Println("===========================================")
	fmt.Printf("Diagnostic for slug: %s\n", slug)
	fmt.Printf("Base URL: %s\n", baseURL)
	fmt.Println("===========================================\n")

	// Test A: GET /models (OpenAI format)
	fmt.Println("--- Test A: GET /models ---")
	testGet(c, baseURL+"/models", apiKey)

	// Test B: Current broken approach
	fmt.Println("\n--- Test B: GET /model/info?" + slug)
	testGet(c, baseURL+"/model/info?"+slug, apiKey)

	// Test C: GET /model/info without filter -> ALL models
	fmt.Println("\n--- Test C: GET /model/info (no params - ALL models) ---")
	testAndParseAllModels(c, baseURL+"/model/info", apiKey, slug)

	// Test D: GET /v1/model/info variant
	fmt.Println("\n--- Test D: GET /v1/model/info?" + slug)
	testGet(c, baseURL+"/v1/model/info?"+slug, apiKey)

	// Test E: check what keys exist in model_info of each returned entry
	fmt.Println("\n--- Check: model_name field format vs slug lookup ---")
	checkModelNameFormat(c, baseURL+"/model/info", apiKey, slug)
}

type liteLLMModelInfo struct {
	ModelName     string                 `json:"model_name"`
	LiteLLMParams map[string]interface{} `json:"litellm_params"`
	ModelInfo     map[string]interface{} `json:"model_info"`
}

func testGet(c *http.Client, fullURL, apiKey string) {
	req, _ := http.NewRequest("GET", fullURL, nil)
	req.Header.Set("Authorization", "Bearer "+apiKey)
	resp, err := c.Do(req)
	if err != nil {
		fmt.Printf("ERROR: %v\n\n", err)
		return
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	fmt.Printf("HTTP %d | bytes=%d\n%s\n\n", resp.StatusCode, len(raw), truncate(string(raw), 800))
}

func testAndParseAllModels(c *http.Client, url, apiKey, targetSlug string) {
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", "Bearer "+apiKey)
	resp, err := c.Do(req)
	if err != nil {
		fmt.Printf("ERROR: %v\n\n", err)
		return
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	body := string(raw)

	fmt.Printf("HTTP %d | total_bytes=%d\n", resp.StatusCode, len(raw))

	var result struct {
		ModelInfo map[string]liteLLMModelInfo `json:"model_info"`
	}
	if err := json.Unmarshal(raw, &result); err != nil {
		fmt.Printf("JSON parse error: %v\n\n", err)
		return
	}

	fmt.Printf("Total models in response: %d\n", len(result.ModelInfo))
	
	found := false
	for key, info := range result.ModelInfo {
		id := ""
		if mi, ok := info.ModelInfo["id"]; ok {
			id = fmt.Sprintf("%v", mi)
		}
		name := info.ModelName
		keyPreview := key
		if len(keyPreview) > 50 {
			keyPreview = keyPreview[:50] + "..."
		}
		fmt.Printf("  key=%s | name=%s | model_info.id=%s\n", 
			keyPreview, name, id)
		
		if containsIgnoreCase(key, targetSlug) || containsIgnoreCase(name, targetSlug) {
			found = true
			fmt.Printf("  >>> MATCH FOR '%s' <<<\n", targetSlug)
			fmt.Printf("    Full key: %s\n", key)
			fmt.Printf("    model_name: %s\n", name)
			fmt.Printf("    model_info.id: %s\n", id)
		}
	}
	if !found {
		fmt.Printf("  >>> NO KEY MATCHED '%s' <<<\n", targetSlug)
		allKeys := make([]string, 0, len(result.ModelInfo))
		i := 0
		for k := range result.ModelInfo {
			if i >= 10 { break }
			allKeys = append(allKeys, k)
			i++
		}
		fmt.Printf("  Sample keys (%d total): %v\n", len(result.ModelInfo), allKeys)
	}
	fmt.Println()
}

func checkModelNameFormat(c *http.Client, url, apiKey, targetSlug string) {
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", "Bearer "+apiKey)
	resp, err := c.Do(req)
	if err != nil || resp.StatusCode != 200 {
		return
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	
	var result struct {
		ModelInfo map[string]liteLLMModelInfo `json:"model_info"`
	}
	if err := json.Unmarshal(raw, &result); err != nil {
		return
	}
	
	for _, info := range result.ModelInfo {
		fmt.Printf("  model_name value: %q \n", info.ModelName)
		fmt.Printf("  model_info keys: ")
		for k := range info.ModelInfo {
			fmt.Printf("%s ", k)
		}
		fmt.Println()
		break // just first one
	}
}

func containsIgnoreCase(a, b string) bool {
	return strings.Contains(strings.ToLower(a), strings.ToLower(b))
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "\n... (truncated)"
}
